go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go

```diff
@@ -0,0 +1,243 @@
+create table customer(
+	customer_id bigint not null auto_increment,
+	email varbinary(128),
+	primary key(customer_id)
+) ENGINE=InnoDB;
+create table corder(
+	order_id bigint not null auto_increment,
+	customer_id bigint,
+	sku varbinary(128),
+	price bigint,
+	primary key(order_id)
+) ENGINE=InnoDB;
+`
+
+	vSchema = `
+{
+	"sharded": true,
+	"vindexes": {
+		"hash": {
+			"type": "hash"
+		}
+	},
+	"tables": {
+		"customer": {
+			"column_vindexes": [
+				{
+					"column": "customer_id",
+					"name": "hash"
+				}
+			]
+		},
+		"corder": {
+			"column_vindexes": [
+				{
+					"column": "customer_id",
+					"name": "hash"
+				}
+			]
+		}
+	}
+}
+`
+)
+
+// TestMain sets up the vitess cluster for any subsequent tests
+func TestMain(m *testing.M) {
+	defer cluster.PanicHandler(nil)
+	flag.Parse()
+
+	exitCode := func() int {
+		clusterInstance = cluster.NewCluster(cell, "localhost")
+		defer clusterInstance.Teardown()
+
+		// Start topo server
+		err := clusterInstance.StartTopo()
+		if err != nil {
+			return 1
+		}
+
+		// Start keyspace
+		keyspace := &cluster.Keyspace{
+			Name:      keyspaceName,
+			SchemaSQL: schemaSQL,
+			VSchema:   vSchema,
+		}
+		clusterInstance.VtTabletExtraArgs = append(clusterInstance.VtTabletExtraArgs, []string{"--health_check_interval", "1s"}...)
+		err = clusterInstance.StartKeyspace(*keyspace, shards, 0, false)
+		if err != nil {
+			return 1
+		}
+
+		clusterInstance.VtGateExtraArgs = append(clusterInstance.VtGateExtraArgs, []string{"--tablet_refresh_interval", tabletRefreshInterval.String()}...)
+		err = clusterInstance.StartVtgate()
+		if err != nil {
+			return 1
+		}
+
+		vtParams = mysql.ConnParams{
+			Host: clusterInstance.Hostname,
+			Port: clusterInstance.VtgateMySQLPort,
+		}
+		return m.Run()
+	}()
+	os.Exit(exitCode)
+}
+
+// TestHealthCheckExternallyReparentNewTablet ensures that calling TabletExternallyReparented on a new tablet will switch the primary tablet
+// without having to wait for the tabletRefreshInterval.
+func TestHealthCheckExternallyReparentNewTablet(t *testing.T) {
+	ctx := context.Background()
+
+	// verify output of `show vitess_tablets` and `INSERT` statement
+	vtgateConn, err := mysql.Connect(ctx, &vtParams)
+	require.Nil(t, err)
+	defer vtgateConn.Close()
+
+	// add a new tablet
+	reparentTabletUID := 9999
+	reparentTabletType := "rdonly"
+	tablet := addTablet(t, reparentTabletUID, reparentTabletType)
+
+	// promote the new tablet to the primary
+	err = clusterInstance.VtctlclientProcess.ExecuteCommand("TabletExternallyReparented", tablet.Alias)
+	require.NoError(t, err)
+
+	// update the new primary tablet to be read-write
+	err = clusterInstance.VtctlclientProcess.ExecuteCommand("SetReadWrite", tablet.Alias)
+	require.NoError(t, err)
+
+	// wait for the vtgate to finish updating the new primary tablet
+	time.Sleep(1 * time.Second)
+
+	// verify that the vtgate will recognize the new primary tablet
+	qr, _ := vtgateConn.ExecuteFetch("show vitess_tablets", 100, true)
```

Comment:
If you run the test code with the slight modifications against the current main, the following output will be generated:

```
...
I0414 16:43:31.320373   87256 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 TabletExternallyReparented healthcheck_test_cell-0000009999 (attempt 1 of 10)
I0414 16:43:31.472102   87256 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 SetReadWrite healthcheck_test_cell-0000009999 (attempt 1 of 10)
    reparent_test.go:161: qr=[[VARCHAR("healthcheck_test_cell") VARCHAR("healthcheck_test_ks") VARCHAR("-80") VARCHAR("REPLICA") VARCHAR("NOT_SERVING") VARCHAR("healthcheck_test_cell-0000001336") VARCHAR("localhost") VARCHAR("")] [VARCHAR("healthcheck_test_cell") VARCHAR("healthcheck_test_ks") VARCHAR("80-") VARCHAR("PRIMARY") VARCHAR("SERVING") VARCHAR("healthcheck_test_cell-0000001337") VARCHAR("localhost") VARCHAR("2023-04-14T07:41:48Z")]]
    reparent_test.go:162:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:162
                Error:          Not equal:
                                expected: 3
                                actual  : 2
                Test:           TestHealthCheckExternallyReparentNewTablet
                Messages:       wrong number of tablet records in healthcheck, expected 3 but had 2
    reparent_test.go:164:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:164
                Error:          Not equal:
                                expected: "PRIMARY"
                                actual  : "REPLICA"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -PRIMARY
                                +REPLICA
                Test:           TestHealthCheckExternallyReparentNewTablet
    reparent_test.go:165:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:165
                Error:          Not equal:
                                expected: "SERVING"
                                actual  : "NOT_SERVING"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -SERVING
                                +NOT_SERVING
                Test:           TestHealthCheckExternallyReparentNewTablet
    reparent_test.go:166:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:166
                Error:          Not equal:
                                expected: "healthcheck_test_cell-0000009999"
                                actual  : "healthcheck_test_cell-0000001336"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -healthcheck_test_cell-0000009999
                                +healthcheck_test_cell-0000001336
                Test:           TestHealthCheckExternallyReparentNewTablet
I0414 16:43:42.524622   87256 mysqlctl_process.go:156] Shutting down MySQL: 1336

I0414 16:43:54.685033   87256 mysqlctl_process.go:175] MySQL shutdown complete: 1336
I0414 16:43:54.764293   87256 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 DeleteTablet healthcheck_test_cell-0000001336 (attempt 1 of 10)
    reparent_test.go:243: Deleted tablet: healthcheck_test_cell-0000001336
    reparent_test.go:175:
                Error Trace:    /Users/yoheimuta/.go/vitess/go/test/endtoend/vtgate/tablet_healthcheck/reparent_test.go:175
                Error:          Received unexpected error:
                                target: healthcheck_test_ks.-80.primary: current keyspace is being resharded (errno 1105) (sqlstate HY000) during query: insert into customer(customer_id, email) values(2, 'dummy1')
                Test:           TestHealthCheckExternallyReparentNewTablet
--- FAIL: TestHealthCheckExternallyReparentNewTablet (93.68s)
```

Reply:
When you run the same test code against the current branch:

```
...
reparent_test.go:225: Added tablet: healthcheck_test_cell-0000009999
I0414 17:05:59.561821   92542 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 TabletExternallyReparented healthcheck_test_cell-0000009999 (attempt 1 of 10)
I0414 17:05:59.755164   92542 vtctlclient_process.go:210] Executing vtctlclient with command: vtctlclient --server localhost:6708 SetReadWrite healthcheck_test_cell-0000009999 (attempt 1 of 10)
    reparent_test.go:161: qr=[ # NOTE: Inserted new lines manually
[VARCHAR("healthcheck_test_cell") VARCHAR("healthcheck_test_ks") VARCHAR("-80") VARCHAR("PRIMARY") VARCHAR("SERVING") VARCHAR("healthcheck_test_cell-0000009999") VARCHAR("localhost") VARCHAR("2023-04-14T08:05:59Z")] 
[VARCHAR("healthcheck_test_cell") VARCHAR("healthcheck_test_ks") VARCHAR("-80") VARCHAR("REPLICA") VARCHAR("NOT_SERVING") VARCHAR("healthcheck_test_cell-0000006955") VARCHAR("localhost") VARCHAR("")] 
[VARCHAR("healthcheck_test_cell") VARCHAR("healthcheck_test_ks") VARCHAR("80-") VARCHAR("PRIMARY") VARCHAR("SERVING") VARCHAR("healthcheck_test_cell-0000006956") VARCHAR("localhost") VARCHAR("2023-04-14T08:04:46Z")]
]
...
```

```
...
I0414 17:05:59.811311   93642 tablet_health_check.go:110] HealthCheckUpdate(Serving State): tablet: healthcheck_test_cell-6955 (localhost) serving true => false for healthcheck_test_ks/-80 (PRIMARY) reason: healthCheck update
I0414 17:05:59.814976   93642 healthcheck.go:509] We will have no health data for the next new primary tablet after demoting the tablet: healthcheck_test_cell-6955 (localhost), so start loading tablets now
I0414 17:05:59.835891   93642 healthcheck.go:382] Adding tablet to healthcheck: alias:{cell:"healthcheck_test_cell" uid:9999} hostname:"localhost" port_map:{key:"grpc" value:6722} port_map:{key:"vt" value:6721} keyspace:"healthcheck_test_ks" shard:"-80" key_range:{end:"\x80"} type:PRIMARY mysql_hostname:"localhost" mysql_port:6723 primary_term_start_time:{seconds:1681459559 nanoseconds:687021000} db_server_version:"5.7.39" default_conn_collation:255
I0414 17:05:59.837507   93642 tablet_health_check.go:110] HealthCheckUpdate(Serving State): tablet: healthcheck_test_cell-9999 (localhost) serving false => true for healthcheck_test_ks/-80 (PRIMARY) reason: healthCheck update
...
```